### PR TITLE
[SPARK-26909][FOLLOWUP][SQL] use unsafeRow.hashCode() as hash value in HashAggregate

### PR DESCRIPTION
**Pull request by yucai**
## What changes were proposed in this pull request?

This is a followup PR for #21149.

New way uses unsafeRow.hashCode() as hash value in HashAggregate.
The unsafe row has [null bit set] etc., so the hash should be different from shuffle hash, and then we don't need a special seed.

## How was this patch tested?

UTs.

### CODE REVIEW

**Comment by yucai** [comment]
Generated codes:

```
/* 054 */     // generate grouping key
/* 055 */     agg_mutableStateArray_0[0].reset();
/* 056 */
/* 057 */     agg_mutableStateArray_0[0].write(0, agg_expr_0_0);
/* 058 */     int agg_unsafeRowHash_0 = (agg_mutableStateArray_0[0].getRow()).hashCode();
/* 059 */     if (true) {
/* 060 */       // try to get the buffer from hash map
/* 061 */       agg_unsafeRowAggBuffer_0 =
/* 062 */       agg_hashMap_0.getAggregationBufferFromUnsafeRow((agg_mutableStateArray_0[0].getRow()), agg_unsafeRowHash_0);
/* 063 */     }
/* 064 */     // Can't allocate buffer from the hash map. Spill the map and fallback to sort-based
/* 065 */     // aggregation after processing all input rows.
/* 066 */     if (agg_unsafeRowAggBuffer_0 == null) {
/* 067 */       if (agg_sorter_0 == null) {
/* 068 */         agg_sorter_0 = agg_hashMap_0.destructAndCreateExternalSorter();
/* 069 */       } else {
/* 070 */         agg_sorter_0.merge(agg_hashMap_0.destructAndCreateExternalSorter());
/* 071 */       }
/* 072 */
/* 073 */       // the hash map had be spilled, it should have enough memory now,
/* 074 */       // try to allocate buffer again.
/* 075 */       agg_unsafeRowAggBuffer_0 = agg_hashMap_0.getAggregationBufferFromUnsafeRow(
/* 076 */         (agg_mutableStateArray_0[0].getRow()), agg_unsafeRowHash_0);
/* 077 */       if (agg_unsafeRowAggBuffer_0 == null) {
/* 078 */         // failed to allocate the first page
/* 079 */         throw new org.apache.spark.memory.SparkOutOfMemoryError("No enough memory for aggregation");
/* 080 */       }
/* 081 */     }
```

**Comment by yucai** [comment]
Before:
```
    val hashExpr = Murmur3Hash(groupingExpressions, 42)
    val hashEval = BindReferences.bindReference(hashExpr, child.output).genCode(ctx)
```


After:

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashAggregateExec.scala [hunk 1]
```diff
@@ -777,11 +771,11 @@ case class HashAggregateExec(
       s"""
          |// generate grouping key
          |${unsafeRowKeyCode.code}
-         |${hashEval.code}
+         |int $unsafeRowHash = ${unsafeRowKeyCode.value}.hashCode();
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashAggregateExec.scala]: nit: `unsafeKeyHash`?

**Review by maropu — APPROVED** [review]
Aha, this is a smarter way...

**Comment by maropu** [comment]
retest this please

**Comment by yucai** [comment]
retest this please

**Comment by yucai** [comment]
@cloud-fan , kindly help review.

**Comment by cloud-fan** [comment]
did you verify it that it can avoid the hash conflict with the problematic query?

**Comment by yucai** [comment]
> did you verify it that it can avoid the hash conflict with the problematic query?

I used another query to verify this issue.

See: https://github.com/apache/spark/pull/23821#issuecomment-464628698

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**